### PR TITLE
fix(server): embed {{prompt}} via shellEscape instead of env-var indirection

### LINKED ISSUE #851: lib/template.ts: replace {{prompt}} env-var indirection with shellEscape direct embedding (privilege-elevation incompatibility)
## Summary

`packages/server/src/lib/template.ts`'s `expandTemplate` replaces `{{prompt}}` with `"$__AGENT_PROMPT__"` and stashes the prompt content in `env.__AGENT_PROMPT__`. This pattern is incompatible with the `sudo -u <user> -i` privilege-elevation path used by `runAsUser` (introduced in Issue [#839](https://github.com/ms2sato/agent-console/issues/839) / PR #840 and consumed by description generation in Issue [#835](https://github.com/ms2sato/agent-console/issues/835) / PR #842). User-visible symptom in multi-user mode after PR #842:

> Description generation failed: Agent command failed: Error: Input must be provided either through stdin or as a prompt argument when using --print
> You can set the description manually from the Settings page.

This Issue replaces the env-var indirection with direct `shellEscape(prompt)` embedding, matching how `{{cwd}}` and custom `{{varName}}` placeholders are already handled. Pre-existing `runAsUser` and PTY worker code paths are unchanged.

## Root cause

Reproduced manually on the dogfood host (Ubuntu 24.04 aarch64, sudo 1.9.15p5):

```
$ sudo -u ms2sato -i sh -c 'cd /tmp && export FOO=bar; pwd; echo "FOO=$FOO"'
/tmp
FOO=
```

`cd` and `export` both run in a coherent inner shell (`pwd` returns `/tmp`), but `$FOO` resolves to empty. Diagnosis:

`sudo -i <command>` forwards `<command>` to the target user's login shell via `-c`. To do so it concatenates and shell-escapes the args, wrapping the inner string in **double quotes**. The login shell (bash) then expands `$VAR` inside the double-quoted arg against its own (empty) environment **before** forking the inner shell. By the time the inner `sh -c` sees the script, all `"$VAR"` references have already been substituted with empty strings.

The PTY worker (`packages/server/src/services/user-mode.ts:478-500`) uses the identical `sudo -u <user> -i sh -c '...'` shape but is unaffected because its inner script only `export`s env vars for inheritance by the spawned process (e.g. `claude`) and never references them via `"$VAR"` in the command itself. The description-generator path (via `expandTemplate`) does both `export` and `"$VAR"` reference — the reference is the broken half.

## Why the env-var pattern is a historical outlier

The `{{prompt}}` env-var indirection predates the `shellEscape` helper:

- `5bd52a3` (PR #108, 2025-12-18) introduced templates: `{{cwd}}` was direct substitution (later flagged as a security gap), `{{prompt}}` was env-var-via-`"$VAR"`.
- `0d8dfc4` (2025-12-19, security review fix) added the `shellEscape` helper and migrated `{{cwd}}` to direct shell-escaped embedding. `{{prompt}}` was not migrated.
- `ec6d2b2` (later) added custom `{{varName}}` template variables — used `shellEscape(value)` for direct embedding, matching the new `{{cwd}}` approach.

Today `{{cwd}}` and `{{varName}}` use `shellEscape` direct embedding, and only `{{prompt}}` retains the env-var pattern. This Issue closes the inconsistency.

## Proposed Change

`packages/server/src/lib/template.ts`:

```ts
// before
if (command.includes('{{prompt}}')) {
  command = command.replace(/\{\{prompt\}\}/g, `"$${PROMPT_ENV_VAR}"`);
  env[PROMPT_ENV_VAR] = prompt ?? '';
}

// after
if (command.includes('{{prompt}}')) {
  command = command.replace(/\{\{prompt\}\}/g, shellEscape(prompt ?? ''));
}
```

Plus comment update explaining the rationale + reference to this Issue.

`shellEscape` (single-quote enclosed, embedded `'` rendered as `'\''`) is POSIX-shell-literal-safe — every metacharacter is neutralised. Same guarantee that `{{cwd}}` and `{{varName}}` already rely on.

`PROMPT_ENV_VAR` constant can be removed entirely (no callers after this change).

## Verification

Reproduced + verified on the same dogfood host:

1. Edited `lib/template.ts` per the proposed change in a separate worktree.
2. Rebuilt `dist/`.
3. Redeployed via `setup-multiuser-for-ubuntu.sh --repo-source <worktree path>`.
4. Triggered description generation through the UI on `Hello-World` repository (cloned from `https://github.com/octocat/Hello-World.git`).
5. The error no longer appears; description is generated successfully.

## Acceptance Criteria

- `expandTemplate` no longer emits `"$__AGENT_PROMPT__"` or populates `env[PROMPT_ENV_VAR]` for the `{{prompt}}` placeholder; instead it emits `shellEscape(prompt)` inline.
- `PROMPT_ENV_VAR` constant removed; no callers remain (grep clean).
- `packages/server/src/lib/__tests__/template.test.ts` updated to assert the new `result.command` shape (direct shell-escaped prompt) and absence of `result.env.__AGENT_PROMPT__`. Existing edge-case coverage (empty prompt, special characters, multiline content, multiple `{{prompt}}` occurrences) preserved.
- Description generation works end-to-end in `AUTH_MODE=multi-user` for the dogfood host scenario (manually verified above; document the steps in the PR body).
- `bun run typecheck`, `bun run test`, `bun run check:lang`, `node .claude/skills/orchestrator/preflight-check.js` all exit 0.

## Out of Scope

- **Refactoring `runAsUser` to forward `opts.env` via `sudo --preserve-env`** so other future callers of shell-expansion-in-elevated-mode are unaffected. Considered; deferred to a separate follow-up if/when a real second caller surfaces. The minimal change in this Issue keeps the helper unchanged and fixes the only known broken consumer.
- **PTY worker (`user-mode.ts`) refactor.** Its inner script does not reference env vars via `"$VAR"` in the command itself; the env-export pattern works correctly for inheritance and is unchanged.
- **Backward compatibility for the env-var pattern.** No external consumer of `expandTemplate`'s return value relies on `env.__AGENT_PROMPT__` being populated (the only consumers are server-internal callers that pass `env` straight to the spawn helper; the spawn becomes a no-op for that key).

## Related

- Issue [#835](https://github.com/ms2sato/agent-console/issues/835) / PR #842 — introduced the description-generator path through `runAsUser` that surfaced this latent bug.
- Issue [#839](https://github.com/ms2sato/agent-console/issues/839) / PR #840 — `runAsUser` helper.
- Umbrella [#837](https://github.com/ms2sato/agent-console/issues/837) — privilege-elevation pattern.
- `packages/server/src/services/user-mode.ts:478-500` — the PTY-worker `sudo -i` reference pattern; works because it does not use `"$VAR"` shell expansion in the inner script.

### PR DESCRIPTION
**Pull request by ms2sato**
## Summary

`expandTemplate` previously replaced `{{prompt}}` with `"$__AGENT_PROMPT__"` and stashed the prompt content in `env.__AGENT_PROMPT__`. This pattern is incompatible with the `sudo -u <user> -i` privilege-elevation path used by `runAsUser` (Issue #839 / PR #840) and exercised by the description generator (Issue #835 / PR #842): sudo wraps the inner command in double quotes when forwarding it to the login shell, which then expands `"$VAR"` against its own (empty) environment before the inner shell sees the command. User-visible symptom in multi-user mode after PR #842 was "Input must be provided either through stdin or as a prompt argument when using --print".

This change replaces the env-var indirection with `shellEscape(prompt ?? '')` direct embedding, matching how `{{cwd}}` and custom `{{varName}}` placeholders are already handled. `PROMPT_ENV_VAR` and `getPromptEnvVar` are removed (no callers remained).

Pre-verified on the dogfood host (Ubuntu 24.04 multi-user mode) in a separate worktree before this PR was opened. See Issue #851 "Verification" section for the manual repro + fix verification trace.

Closes #851

## Files changed

- `packages/server/src/lib/template.ts` -- production fix: `{{prompt}}` now expands via `shellEscape` direct embedding. `PROMPT_ENV_VAR` constant and `getPromptEnvVar` exported function removed.
- `packages/server/src/lib/__tests__/template.test.ts` -- every `result.command` assertion updated to expect the shell-escaped inline prompt; every `result.env.__AGENT_PROMPT__` assertion removed; `getPromptEnvVar` describe block removed; edge-case coverage (empty / shell-meta / quotes / newlines / dollar signs / paths) preserved.
- `packages/server/src/__tests__/utils/extract-prompt-from-command.ts` -- new shared test helper that extracts the prompt from a spawn command string (parses the trailing single-quoted segment, reverses the `'\''` escape).
- `packages/server/src/services/__tests__/repository-description-generator.test.ts` -- three spawn-call assertions (README content, prompt instruction, output format) and the elevated-branch inner-command assertion updated to extract the prompt from `args[2]` / inner via the helper; explicit assertion that `env.__AGENT_PROMPT__` is `undefined` in the non-elevated branch and not exported in the elevated inner command.
- `packages/server/src/services/__tests__/session-metadata-suggester.test.ts` -- same pattern; existingBranches / no-branches cases use the helper.
- `packages/server/src/mcp/__tests__/mcp-server.test.ts` -- `getAgentPromptForSession` helper updated to extract the prompt from the PTY spawn args[1][1] instead of options.env.

## Verification log

```
$ bun run typecheck
@agent-console/shared typecheck: Exited with code 0
@agent-console/server typecheck: Exited with code 0
@agent-console/client typecheck: vite v6.4.1 building for production...
@agent-console/client typecheck: transforming...
@agent-console/client typecheck: ✓ 689 modules transformed.
@agent-console/client typecheck: ✓ built in 2.42s
@agent-console/client typecheck: Exited with code 0
```

```
$ output=$(bun run test 2>&1); exitcode=$?; echo "$output" | tail -100; echo "TEST_EXIT: $exitcode"
[... full last 100 lines of test output, last block reproduced below ...]
@agent-console/client test:  1397 pass
@agent-console/client test:  2 skip
@agent-console/client test:  0 fail
@agent-console/client test:  2591 expect() calls
@agent-console/client test: Ran 1399 tests across 88 files. [15.28s]
@agent-console/client test: Exited with code 0
$ bun test scripts/ ./.claude/hooks/__tests__/ ./.claude/skills/*/__tests__/
bun test v1.3.10 (30e609e0)
 362 pass
 0 fail
 896 expect() calls
Ran 362 tests across 11 files. [2.42s]
TEST_EXIT: 0
```

```
$ bun run check:lang
OK -- language check clean (103 files scanned).
```

```
$ node .claude/skills/orchestrator/preflight-check.js
## Test Coverage Check
No production files matching coverage patterns were changed.
## Rule/Skill Duplication Check
No rule paragraphs found verbatim in any skill file.
## Language Check (public artifacts)
All public artifacts use Latin / Greek / Cyrillic scripts only.
```

(Note: preflight reports "No production files matching coverage patterns were changed" because the touched production file `packages/server/src/lib/template.ts` falls under `lib/` rather than the routes/services/components paths listed in `COVERAGE_PATTERNS`. The sibling test `packages/server/src/lib/__tests__/template.test.ts` is still updated in this PR.)

## TDD polarity check

Stashed `packages/server/src/lib/template.ts` (the production fix only, leaving the updated test file in place), then re-ran `bun test src/lib/__tests__/template.test.ts`:

```
 9 pass
 26 fail
 38 expect() calls
Ran 35 tests across 1 file. [13.00ms]
```

Sample failure shape:

```
Expected: "test-cli 'Hello World'"
Received: "test-cli "$__AGENT_PROMPT__""
```

Restored the fix; all 35 template tests pass again. The new assertions genuinely verify the new shellEscape-embed behaviour and would fail on a future regression to the env-var pattern.

## 7-question pre-PR gap-scan

1. **Existing mechanism?** Yes -- `shellEscape` is already used for `{{cwd}}` (line 64) and custom `{{varName}}` (line 82). This PR aligns `{{prompt}}` with the existing pattern. No duplication.
2. **Trigger / invocation documented?** N/A -- internal refactor; `expandTemplate`'s public contract (return shape `{ command, env }`) is unchanged from callers' perspectives. The `env` field is now empty for prompt callers but remains available for future placeholders that may need env indirection.
3. **Failure paths tested?** Yes -- empty prompt, undefined prompt, shell metacharacters, quotes (single + double), newlines, dollar signs, paths with spaces / quotes, multiple `{{prompt}}` occurrences. All preserved from the existing edge-case set. Sibling tests (`template.test.ts`, the three consumer test files) all updated in this PR.
4. **New file type / lifecycle?** No -- the only new file is a test utility (`extract-prompt-from-command.ts`) under the existing `packages/server/src/__tests__/utils/` directory.
5. **Rule clarity?** N/A -- no rule changes.
6. **Cross-runtime spawn?** No -- no new spawn boundary; `runAsUser` continues to dispatch to `Bun.spawn` as before.
7. **Shared / persistent artifact write?** No -- no installer / daemon / hook / package metadata changes.
8. **Signature shape change?** No -- `expandTemplate`'s signature and return shape are unchanged. The `getPromptEnvVar` exported helper is removed; only one consumer (the test file) referenced it, and that consumer is updated in the same PR.

## Out of scope (per Issue #851)

- Refactoring `runAsUser` to forward `opts.env` via `sudo --preserve-env` (deferred unless a real second caller surfaces).
- PTY worker (`user-mode.ts`) refactor -- its inner script does not reference env vars via `"$VAR"` in the command itself; the env-export pattern works correctly for inheritance and remains unchanged.
- Backward compatibility for the env-var pattern -- no external consumer relies on `env.__AGENT_PROMPT__` being populated.

### CODE REVIEW

File: packages/server/src/mcp/__tests__/mcp-server.test.ts [hunk 1]
```diff
@@ -1806,18 +1807,23 @@ describe('MCP Server Tools', () => {
...
     /**
-     * Extract the __AGENT_PROMPT__ env var from the PTY spawn call
-     * that matches the given session ID.
+     * Extract the embedded prompt from the PTY spawn call that matches the
+     * given session ID. After Issue #851, the prompt is embedded directly
+     * into the spawn command via shellEscape (single-quoted literal),
+     * instead of being indirected through env.__AGENT_PROMPT__.
      */
     function getAgentPromptForSession(sessionId: string): string {
       const calls = ptyFactory.spawn.mock.calls as unknown as Array<[string, string[], PtySpawnOptions]>;
```
[the remark this answers is not in the document]
Comment (ms2sato) [reply packages/server/src/mcp/__tests__/mcp-server.test.ts:1823]: The direct tuple cast fails with TS2352 in strict mode:

```
error TS2352: Conversion of type '[][]' to type '[string, string[], PtySpawnOptions | undefined][]' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
```

The root cause is that `bun:test`'s `mock(() => ...)` (defined in `packages/server/src/__tests__/utils/mock-pty.ts:96`) has no declared parameters, so `mock.calls` is inferred as `[][]`. TypeScript itself instructs to "convert the expression to 'unknown' first", which is exactly what the existing `as unknown as` does.

The same constraint applies to the two pre-existing sibling casts in this file at line 1311 (`findSpawnCallByCommand`) and line 2378 (activity-detector probe).

Reverted the direct-cast attempt in 4fad085 and added an inline comment explaining the TS constraint so future readers (and CR) do not re-walk this path. The cast remains consistent with the two pre-existing instances in the same file.

Out of scope for this PR (which is narrowly Issue #851): refactoring `createMockPtyFactory` to give `spawn` a typed signature would let the direct cast work and let us remove all three `as unknown as` uses in this file. That would be the right place to address the standards rule on `unknown as`. Happy to file a follow-up if you'd like.
